src/dstack/plugins/builtin/rest_plugin.py

```
@@ -0,0 +1,97 @@
+from typing import Generic, TypeVar
+
+import requests
+from pydantic import BaseModel, ValidationError
+
+from dstack._internal.core.errors import ServerClientError
+from dstack._internal.core.models.fleets import FleetSpec
+from dstack._internal.core.models.gateways import GatewaySpec
+from dstack._internal.core.models.volumes import VolumeSpec
+from dstack.plugins import ApplyPolicy, Plugin, RunSpec, get_plugin_logger
+from dstack.plugins._models import ApplySpec
+
+logger = get_plugin_logger(__name__)
+
+PLUGIN_SERVICE_URI_ENV_VAR_NAME = "DSTACK_PLUGIN_SERVICE_URI"
+PLUGIN_REQUEST_TIMEOUT = 8  # in seconds
+
+SpecType = TypeVar("SpecType", RunSpec, FleetSpec, VolumeSpec, GatewaySpec)
+
+
+class SpecRequest(BaseModel, Generic[SpecType]):
+    user: str
+    project: str
+    spec: SpecType
+
+
+RunSpecRequest = SpecRequest[RunSpec]
+FleetSpecRequest = SpecRequest[FleetSpec]
+VolumeSpecRequest = SpecRequest[VolumeSpec]
+GatewaySpecRequest = SpecRequest[GatewaySpec]
+
+
+class CustomApplyPolicy(ApplyPolicy):
+    def __init__(self):
+        self._plugin_service_uri = os.getenv(PLUGIN_SERVICE_URI_ENV_VAR_NAME)
+        logger.info(f"Found plugin service at {self._plugin_service_uri}")
+        if not self._plugin_service_uri:
+            logger.error(
+                f"Cannot create policy because {PLUGIN_SERVICE_URI_ENV_VAR_NAME} is not set"
+            )
+            raise ServerClientError(f"{PLUGIN_SERVICE_URI_ENV_VAR_NAME} is not set")
+
+    def _call_plugin_service(self, spec_request: SpecRequest, endpoint: str) -> ApplySpec:
+        response = None
+        try:
+            response = requests.post(
+                f"{self._plugin_service_uri}{endpoint}",
+                json=spec_request.dict(),
+                headers={"accept": "application/json", "Content-Type": "application/json"},
+                timeout=PLUGIN_REQUEST_TIMEOUT,
+            )
+            response.raise_for_status()
+            spec_json = json.loads(response.text)
```

How does the Plugin reject a Run from being applied if we swallow this exception? Do we want some exceptions (HTTP Status Codes) to be propagated to Dstack? Or should there be an `errors` element of the response body?

I'm leaning toward the latter since if we're monitoring the plugin server, it might be difficult to distinguish between "the plugin server thought this request was invalid" and "the plugin server could not answer the question".

---

Sorry, I misread your comment when you shared it internally. Yes - I agree on the latter option, that way the user knows if the action is on their end or if it's a system issue (a bug in dstack or plugin server)

---

Added a pydantic model for the plugin server response that includes an `error` field [here](https://github.com/dstackai/dstack/pull/2631/commits/e757a77789072cf8c810e9c47e60bc1e9e396d84#diff-cefd5b27016f62b497ed8ccba936cd08c172b06b328ae446e99d600ab1982de3R32). Expectation is if request is rejected, dstack should a 200 response from plugin server with an `error`.